core/rawdb/freezer_batch.go

```diff
@@ -182,13 +188,108 @@ func (batch *freezerBatch) write(newHead bool) (bool, error) {
 	if batch.count > 0 {
 		// Some data left to write on a retry.
 		batch.sizes = batch.sizes[count:]
+		batch.total = 0
+		for _, size := range batch.sizes {
+			batch.total += int(size)
+		}
 		return true, nil
 	}
 	// All data written. We can simply truncate and keep using the buffer
 	batch.sizes = batch.sizes[:0]
+	batch.total = 0
 	return false, nil
 }
 
-func (batch *freezerBatch) Size() int {
-	return batch.buf.Len()
+func (batch *freezerTableBatch) Size() int {
+	return batch.total
+}
+
+// freezerBatch is a write-only database that commits changes to the associated
+// ancient store when Write is called. A freezerBatch cannot be used concurrently.
+type freezerBatch struct {
+	lock    sync.RWMutex
```

Comment:
The docs say it cannot be used concurrently -- so why the lock?

Reply:
The lock is not `Reset`. The reset will "write" the batch map while all other operation will just "read" the batch map. Just safer to add the lock here. But if there is a big performance difference, than we can take the risk to remove the lock here.